apps/cassandra.go

```diff
@@ -0,0 +1,206 @@
+					// 				org.apache.cassandra.io.FSReadError: java.io.IOException: Invalid folder descriptor trying to create log replica /folder/views-9786ac1cdd583201a7cdad556410c985
+					// 					at org.apache.cassandra.db.lifecycle.LogReplica.create(LogReplica.java:59)
+					// 					at org.apache.cassandra.db.lifecycle.LogReplicaSet.maybeCreateReplica(LogReplicaSet.java:87)
+					// 					at org.apache.cassandra.db.lifecycle.LogFile.makeAddRecord(LogFile.java:336)
+					// 					at org.apache.cassandra.db.lifecycle.LogFile.add(LogFile.java:310)
+					Regex: `^(?<level>[A-Z]+)\s+\[(?<module>[^\]]+)\]\s+(?<time>\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2},\d+)\s+(?<message>(?:(?<javaClass>[\w\.]+):(?<lineNumber>\d+))?[\S\s]+)`,
+					Parser: confgenerator.ParserShared{
+						TimeKey:    "time",
+						TimeFormat: "%Y-%m-%d %H:%M:%S,%L",
+						Types: map[string]string{
+							"lineNumber": "integer",
+						},
+					},
+				},
+			},
+		},
+		Rules: []confgenerator.MultilineRule{
+			confgenerator.MultilineRule{
+				StateName: "start_state",
+				NextState: "cont",
+				Regex:     `[A-Z]+\s+\[[^\]]+\] \d+`,
+			},
+			confgenerator.MultilineRule{
+				StateName: "cont",
+				NextState: "cont",
+				Regex:     `^(?![A-Z]+\s+\[[^\]]+\] \d+)`,
+			},
+		},
+	}.Components(tag, uid)
+
+	c = append(c,
+		fluentbit.TranslationComponents(tag, "level", "logging.googleapis.com/severity",
+			[]struct{ SrcVal, DestVal string }{
+				{"TRACE", "TRACE"},
+				{"DEBUG", "DEBUG"},
+				{"INFO", "INFO"},
+				{"ERROR", "ERROR"},
+				{"WARN", "WARNING"},
+			},
+		)...,
+	)
+
+	return c
+}
+
+type LoggingProcessorCassandraGC struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorCassandraGC) Type() string {
+	return "cassandra_gc"
+}
+
+func (p LoggingProcessorCassandraGC) Components(tag string, uid string) []fluentbit.Component {
+	c := confgenerator.LoggingProcessorParseMultilineRegex{
+		LoggingProcessorParseRegexComplex: confgenerator.LoggingProcessorParseRegexComplex{
+			Parsers: []confgenerator.RegexParser{
+				confgenerator.RegexParser{
+					// Vast majority of lines look like the first, with time stopped & time stopping
+					// Sample line: 2021-10-02T04:18:28.284+0000: 3.315: Total time for which application threads were stopped: 0.0002390 seconds, Stopping threads took: 0.0000281 seconds
+					// Sample line: 2021-10-05T01:20:52.695+0000: 4.434: [GC (CMS Initial Mark) [1 CMS-initial-mark: 0K(3686400K)] 36082K(4055040K), 0.0130057 secs] [Times: user=0.04 sys=0.00, real=0.01 secs]
+					// Sample line: 2021-10-05T01:20:52.741+0000: 4.481: [CMS-concurrent-preclean-start]
+					// Lines may also contain more detailed GC Heap information in the following lines
+					Regex: `^(?<time>\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3,6}(?:Z|[+-]\d{2}:?\d{2})):\s+(?<uptime>\d+\.\d{3,6}):\s+(?<message>(?:.*(?:stopped: (?<timeStopped>\d+\.\d+)).*(?:took: (?<timeStopping>\d+\.\d+)[\s\S]*)|[\s\S]+))`,
```

Comment:
Can we make the regex for timeStopped and timeStopping more strict to avoid accidentally matching something else and hopefully optimize the regex performance a bit?

e.g.

Match `Total time for which application threads were stopped` instead of `stopped`, and match `Stopping threads took` instead of `took` and match ` seconds $` (end of line). Does that have any down side?

Reply:
Cool. Resolving this.